### PR TITLE
[MM-25355] Throttle notifications for Windows by channel id

### PR DESCRIPTION
**Pull request by devinbinnie**
**Summary**
On Windows, when receiving multiple notifications sequentially, the application would show them in order but there would be a delay between each notification and if there was a large number of notifications, the application would spam the user.

This PR stops the spam of notifications by only storing the last notification from each channel id in the Notifications Center and auto-dismissing all previous notifications.

**Issue link**
https://mattermost.atlassian.net/browse/MM-25355

### CODE REVIEW

**Comment by devinbinnie** [comment]
Notes:
- @esethna Added you as reviewer to confirm that this change fixes the issue in your eyes.
- Includes a similar change to https://github.com/mattermost/desktop/pull/1484, which I'll merge into this branch once that PR is merged.

**Comment by esethna** [comment]
@devinbinnie looks like build failed

**Comment by devinbinnie** [comment]
> @devinbinnie looks like build failed

Build always fails right now, I think there's an issue with the CI. Just wanted your take on the approach as described above.
There will be a build you can test out if this gets merged.

File: src/main/notifications/index.js
Comment (hmhealey) [inline src/main/notifications/index.js:35]: Instead of using a `Set` and iterating over it, you could use a `Map` mapping channel IDs to pending notifications instead. 0/5 on that though

File: src/main/notifications/index.js
Comment (hmhealey) [inline src/main/notifications/index.js:38]: Do we need to do anything to remove notifications from `currentNotifications` when they're dismissed?

**Review by Willyfrog — CHANGES_REQUESTED** [review]
some thoughts:
- we could listen to the 'close' event (which is not guaranteed but it happens most of the time) to remove the channel from the queue to allow new notifications to arrive even if the user is not doing anything.
- since that event is not guaranteed, I'd also add a timer to remove it after a while, otherwise we might get some reports about missing notifications.
- also, I would recomment usign the tabName+team+channel as the key to prevent having two channels with the same name on different servers/teams obscuring each other.

File: src/main/notifications/index.js
Comment (devinbinnie) [reply src/main/notifications/index.js:38]: Isn't that what `currentNotifications.delete` is doing? Should be removing it from the `Set`. Do you mean destroying the object?

File: src/main/notifications/index.js
Comment (Willyfrog) [reply src/main/notifications/index.js:35]: I agree with @hmhealey using a composite key in a map makes more sense, as we want to have a unique element for the key, but the value will be different.

File: src/main/notifications/index.js
Comment (Willyfrog) [inline src/main/notifications/index.js:38]: if the notification was closed by user action, will a second close on it fail somehow?

**Review by Willyfrog — APPROVED** [review]
as it is, should work ok, added a couple of comments for consideration :)

**Comment by esethna** [comment]
> This PR stops the spam of notifications by only storing the last notification from each channel id in the Notifications Center and auto-dismissing all previous notifications.

@devinbinnie This only applies to Windows correct? I think this approach is fine but will need thorough testing cc// @jgilliam17

File: src/main/notifications/index.js
Comment (esethna) [reply src/main/notifications/index.js:38]: @jgilliam17 on this test case ^

**Comment by devinbinnie** [comment]
> > This PR stops the spam of notifications by only storing the last notification from each channel id in the Notifications Center and auto-dismissing all previous notifications.
> 
> @devinbinnie This only applies to Windows correct? I think this approach is fine but will need thorough testing cc// @jgilliam17

Correct, and agreed on the testing.

**Comment by jgilliam17** [comment]
@devinbinnie Please give me a ping here or in a DM to let me know which desktop daily build this is going to be on. Thanks.

File: src/main/notifications/index.js
Comment (hmhealey) [reply src/main/notifications/index.js:38]: I mean, is there anything to do if the notification is dismissed by the user before this triggers? I think that's the same thing Guillermo is asking about as well.

Comment (devinbinnie) [reply src/main/notifications/index.js:38]: Nope, if it's dismissed then `close()` just doesn't do anything. In terms of the `Set` we always just get rid of the last notification regardless.

**Review by hmhealey — APPROVED** [review]

**Comment by Willyfrog** [comment]
@jgilliam17 this should be in the daily build already

**Pull request merged**
